feat: Introduce a 'slot no longer available' message

hariombalhara opened the pull request:
## What does this PR do?
- Ensures that unavailable slot isn't bookable. 'Confirm' button is disabled
[Demo](https://www.loom.com/share/a61590231ee34131b78f7c9a65cda075)
[Previous implementation Loom - UI changed after that](https://www.loom.com/share/c116e3dda86649c8bb0c360f44baaba8)

[Minimum Booking Notice enforcement](https://www.loom.com/share/c92506bdf39e442e9d71fe13baa46dca)


## Fixes in reservation logic. 
- It was considering the slot to be unavailable even for the booker for which it was reserved. Looks like a day zero implementation bug. 
- A slot could be reserved by multiple people which didn't make sense. 


## Deployment Plan
Add env variable. This is required to be set, other env variables have good defaults.
- `PUBLIC_INVALIDATE_AVAILABLE_SLOTS_ON_BOOKING_FORM=1` - Setting it would send more `getSchedule` calls but only on going from slots -> form or vice versa.

## Followup
- Add a client side check as well in quick availability check. Client side check could be for minimum booking notice. That would make the confirm button go disabled immediately.
- Team Event's slot reservation logic isn't working. It doesn't consider the assignees at all. We need to figure out a way to solve it for Collective and RR both but with different approaches.
 

## Mandatory Tasks (DO NOT REMOVE)
- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
See loom
- Test Skip confirm flow as well as that has quite different handling.

Commit:
feat: Don't allow unavailable slot's booking form to be visible - Send back to slots listing

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=18935) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=18935) before we can accept your contribution.<br/><hr/>**Hariom Balhara** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=18935) it.</sub>

hariombalhara commented on apps/web/test/lib/getSchedule/selectedSlots.test.ts:
Reservation flow tests in getSchedule

hariombalhara commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -90,6 +103,52 @@ async function getEventTypeId({
+
+  await _cleanupSlots({ slotsSelectedByTheBooker, eventTypeId });
+
+  const reservedSlots = slotsSelectedByOtherUsers;
```
Change: We must consider these as reservedSlots

hariombalhara commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -90,6 +103,52 @@ async function getEventTypeId({
+  const slotsSelectedByTheBooker = unexpiredSelectedSlots.filter((slot) => slot.uid === bookerClientUid);
+  const slotsSelectedByOtherUsers = unexpiredSelectedSlots.filter((slot) => slot.uid !== bookerClientUid);
+
+  await _cleanupSlots({ slotsSelectedByTheBooker, eventTypeId });
```
Cleanup should clean slots by this booker

hariombalhara commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -90,6 +103,44 @@ async function getEventTypeId({
+    await prisma.selectedSlots.deleteMany({
+      where: {
+        eventTypeId: { equals: eventTypeId },
+        releaseAt: { lt: currentTimeInUtc },
```
Change: We were earlier cleaning up other users selected slots as well. Now we clean this eventType's expired slots only.

I think we could even remove eventTypeId condition, but deletion could be slow query and don't want to slow down availability request

hariombalhara commented on packages/features/bookings/Booker/Booker.tsx:
```diff
@@ -152,12 +152,48 @@ const BookerComponent = ({
+   * @param params.dateString - The date string in YYYY-MM-DD format
+   * @returns boolean - true if the slot is available, false otherwise
+   */
+  const isTimeSlotAvailable = ({
```
This function's logic is quite critical as accidentally returning false from here could prevent users from booking.

hariombalhara commented on packages/features/bookings/Booker/Booker.tsx:
```diff
@@ -152,12 +152,48 @@ const BookerComponent = ({
+    // Check if the exact time slot exists in the available slots
+    return slotsForDateInUtc.some((slot) => {
+      const slotTimeInUtc = slot.time;
+      return slotTimeInUtc === slotToCheckInUTC;
```
We know that the slots returned by `getSchedule` are already in UTC and slotToCheck is also in UTC. So this works.

Commit:
Merge remote-tracking branch 'origin/main' into dont-allow-booking-form-with-unavaialble-slot

Commit:
Improvements

Commit:
feat: Add env variables and strategies to prevent booking failures

- Added new environment variables for configuring slot reservation and availability checks
- Implemented strategies to handle concurrent bookings and slot availability
- Updated Booker component and related files to support new reservation and slot checking mechanisms
- Added README with detailed explanation of booking prevention strategies
- Configured dynamic intervals for slot and reservation queries

hariombalhara commented on packages/features/bookings/Booker/components/hooks/useSlots.ts:
```diff
@@ -36,15 +40,37 @@ export const useSlots = (event: { data?: Pick<BookerEvent, "id" | "length"> | nu
+
   const handleReserveSlot = () => {
-    if (event?.data?.id && selectedTimeslot && (selectedDuration || event?.data?.length)) {
+    if (eventTypeId && slotUtcStartDate && slotUtcEndDate) {
       reserveSlotMutation.mutate({
-        slotUtcStartDate: dayjs(selectedTimeslot).utc().format(),
-        eventTypeId: event.data.id,
-        slotUtcEndDate: dayjs(selectedTimeslot)
-          .utc()
-          .add(selectedDuration || event.data.length, "minutes")
-          .format(),
```
Simplified and moved to variables

Commit:
docs: Update Booker README with detailed slot reservation and availability explanation

- Clarified slot reservation behavior when multiple users access the same booking page
- Added details about `getSchedule` refetching strategies and timing
- Explained how slot availability is detected and communicated to users

Commit:
Merge remote-tracking branch 'origin/main' into dont-allow-booking-form-with-unavaialble-slot

Commit:
Support for skip confirmation form flow

Commit:
feat: Enhance slot availability and reservation mechanism

- Refactored slot availability checking to support multiple slots
- Added tentative selected timeslots to improve booking UX
- Updated trpc handler to check availability for multiple slots
- Introduced new store methods for managing tentative slot selections
- Improved slot reservation and availability status tracking

Commit:
refactor: Improve slot availability quick check mechanism

- Renamed and restructured slot availability check parameters
- Extracted quick availability checks into a separate hook
- Updated type definitions for slot status and availability checks
- Simplified slot availability logic in Booker and related components
- Maintained cached state for quick availability checks

Udit-takkar committed:
Merge branch 'main' into dont-allow-booking-form-with-unavaialble-slot

Udit-takkar reviewed (commented):
There is a bug that sometimes the meeting goes through even when it first displayed 'Unavailable timeslot' error. It's difficult to reproduce it again.


https://github.com/user-attachments/assets/17476574-75a2-4b85-91bc-86d79af38c1c

2) when you click on 'Please select a new time' we should refetch the slots to remove the busy slots.

hariombalhara commented:
I think 1 can happen if 'getSchedule' call hasn't completed ( which is fetched when tab is focussed) and the other tab doesn't have the context. We should be fine for now.

Yeah, 2 makes sense.

Commit:
Merge remote-tracking branch 'origin/main' into dont-allow-booking-form-with-unavaialble-slot

alishaz-polymath reviewed (commented):
@hariombalhara Let's add the minimumBookingNotice check as part of isAvailable here, as we discussed yesterday 🙏

Commit:
Support minimu booking check too

Commit:
Merge branch 'dont-allow-booking-form-with-unavaialble-slot' of https://github.com/calcom/cal.com into dont-allow-booking-form-with-unavaialble-slot

hariombalhara commented:
@alishaz-polymath Done

hariombalhara commented:
@Udit-takkar 2nd point is already happening.

So, it could be that you don't have PUBLIC_INVALIDATE_AVAILABLE_SLOTS_ON_BOOKING_FORM=1 set in env or `getSchedule` call just didn't complete

Commit:
change data-testid

alishaz-polymath commented on packages/lib/isOutOfBounds.tsx:
```diff
@@ -226,6 +227,37 @@ export function isTimeOutOfBounds({
+ * Wrapper over isTimeOutOfBounds to return a status object.
+ * Note: It doesn't throw any error and can be safely used
+ */
+export function getTimeOutOfBoundsStatus({
+  time,
+  minimumBookingNotice,
+}: {
+  time: dayjs.ConfigType;
+  minimumBookingNotice?: number;
+}): {
+  isOutOfBounds: boolean;
+  reason: "minBookNoticeViolation" | "slotInPast" | null;
+} {
+  try {
+    const isOutOfBounds = isTimeOutOfBounds({ time, minimumBookingNotice });
+    return {
+      isOutOfBounds,
+      reason: isOutOfBounds ? "minBookNoticeViolation" : null,
+    };
+  } catch (error) {
+    if (error instanceof BookingDateInPastError) {
+      return {
+        isOutOfBounds: true,
+        reason: "slotInPast",
+      };
+    }
+    throw error;
+  }
+}
```
Technically, minimumBookingNotice is only one of the reasons why we might face`booking_time_out_of_bounds_error`, so for clarity purposes I'd recommend we either rename this check or include check for the other factors that might lead to the said error.

The other factors are when the user has “limit future bookings” selected,
- Rolling: This would fail if the booking slot selected is far into the future (beyond 30 days, for example when 30 days is selected here)
- Rolling-Window: This would fail if the booking slot selected is far into the future (beyond n days of “always ‘n’ days available”)
- Range: This would fail if the booking slot selected is outside of the chosen range

hariombalhara replied:
Renamed

Commit:
Rneame

hariombalhara committed:
Fix race condition with uid cookie not set

hariombalhara commented:
Been trying to fix these duration limits tests. No idea why they are failing consistently. 

Locally they fail for me even in main branch.
I am OOO tomorrow,  but will try to fix these tests so that it lands in main soon.

Commit:
fix tests

hariombalhara commented:
All tests passing now.

hariombalhara commented on apps/web/playwright/booking-limits.e2e.ts:
```diff
@@ -266,9 +269,8 @@ test.describe("Booking limits", () => {
-        await bookTimeSlot(page, { expectedStatusCode: 401 });
 
-        await expect(page.getByTestId("booking-fail")).toBeVisible({ timeout: 5000 });
+        await expectSlotNotAllowedToBook(page);
```
Now, booking isn't possible(instead of a failure earlier) so we use different approach to verify it.

Commit:
Remove unsded variable

alishaz-polymath reviewed (dismissed):
Tested and it works for most part! Great work!! 🚀 🚀 🚀 
There is still a window where the user might be able to click the confirm button despite crossing to the minimumBookingNotice window, depending on the polling time (refetchInterval) and when they landed on the booking form page. 

I think since we already know about the minimumBookingNotice, and it's a static information, we could in theory use that to disable the confirm button at precisely the right time, and that could be a nice follow up strategy to cover these cases.

ThyMinimalDev commented:
testing with atoms

ThyMinimalDev committed:
Merge branch 'main' into dont-allow-booking-form-with-unavaialble-slot

ThyMinimalDev commented on packages/features/bookings/Booker/Booker.tsx:
```diff
@@ -179,6 +220,15 @@ const BookerComponent = ({
     return setBookerState("booking");
   }, [event, selectedDate, selectedTimeslot, setBookerState, skipConfirmStep]);
 
+  const unavailableTimeSlots = allSelectedTimeslots.filter((slot) => {
```
allSelectedTimeslots is not given by the BookerPlatformWrapper which cause an error

ThyMinimalDev replied:
we need to update packages/platform/atoms/hooks/useSlots.ts to return this data, we can return empty data until we implement the feature on api v2

ThyMinimalDev replied:
ill push a commit for it

hariombalhara replied:
Thanks

ThyMinimalDev committed:
fixup! Merge branch 'main' into dont-allow-booking-form-with-unavaialble-slot

CarinaWolli commented:
I just had the slot no longer available error, but it should have been still available. When I clicked back the slot was still there, clicking on the slot again also triggered the error again. 

That's my settings: 
- 1 min event with 1 min before event buffer 
- at 1:15 I clicked on the 1:17 slot
- Error showed at 1:15 --> should be still available 
- When going back to the slots, 1:17 was still there

Commit:
Merge remote-tracking branch 'origin/main' into dont-allow-booking-form-with-unavaialble-slot

hariombalhara commented:
> > I just had the slot no longer available error, but it should have been still available. When I clicked back the slot was still there, clicking on the slot again also triggered the error again.
> > That's my settings:
> > 
> > * 1 min event with 1 min before event buffer
> > * at 1:15 I clicked on the 1:17 slot
> > * Error showed at 1:15 --> should be still available
> > * When going back to the slots, 1:17 was still there
> 
> @CarinaWolli One thing that we do to confirm in this case, is to remove `disabled` property from the confirm button(using inspect element) and let the booking go through. If it passes, then it is a bug but if it fails then I think not a bug.
> 
> Trying to replicate the issue now.

Also, `isAvailable` endpoint returns the reason for unavailability too, so that could tell why it was blocked.

Commit:
Fix the unavailability bug when only seconds differ

hariombalhara commented:
I have added a fix to handle this weird scenario @CarinaWolli. Please test again

CarinaWolli reviewed (commented):
Looks good from side, let's wait for @emrysal's review

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into dont-allow-booking-form-with-unavaialble-slot

hariombalhara committed:
Handle column view that doesnt have selected date and also add tests for isTimeSlotAavailable

emrysal reviewed (approved):
LGTM, happy with this and works great

keithwillcode committed:
Merge branch 'main' into dont-allow-booking-form-with-unavaialble-slot

ThyMinimalDev reviewed (approved):
all good on atoms side, we will need to implement the feature for api v2 and platform in the future
